game-core/src/main/java/games/strategy/triplea/util/MovableUnitsFilter.java

```diff
@@ -0,0 +1,205 @@
+package games.strategy.triplea.util;
+
+import com.google.common.base.Preconditions;
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.MoveDescription;
+import games.strategy.engine.data.PlayerId;
+import games.strategy.engine.data.Route;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.data.UnitCollection;
+import games.strategy.triplea.delegate.AbstractMoveDelegate;
+import games.strategy.triplea.delegate.AbstractMoveDelegate.MoveType;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.MoveValidator;
+import games.strategy.triplea.delegate.UndoableMove;
+import games.strategy.triplea.delegate.UnitComparator;
+import games.strategy.triplea.delegate.data.MoveValidationResult;
+import games.strategy.triplea.delegate.data.MustMoveWithDetails;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import org.triplea.java.collections.CollectionUtils;
+
+/**
+ * Utility class for filtering a unit list to the subset of those units that can move on a route.
+ */
+public final class MovableUnitsFilter {
+  /** The result of the filter operation. */
+  @Getter
+  public static class Result {
+    // The filtered units and dependents that can move on the route.
+    private final List<Unit> unitsWithDependents;
+    // An error message, if no units could move on the route.
+    @Nullable private String errorMessage;
+    // A warning message, if some, but not all units could move on the route.
+    @Nullable private String warningMessage;
+
+    public Result(
+        final MoveValidationResult allUnitsResult,
+        final MoveValidationResultWithDependents lastResult) {
+      this.unitsWithDependents = lastResult.getUnitsWithDependents();
+      if (!allUnitsResult.isMoveValid()) {
+        final @Nullable final String message = getErrorOrWarningMessage(allUnitsResult);
```

Comment:
Interesting that: `final @Nullable final` compiles :grin: 

AFAIK nullable is not used on local variables as the scope is pretty well managed. Recommend to use Optional to call this out.

```
getErrorOrWarningMessage(allUnitsResult)
   .ifPresent(msg -> 
              if( !lastResult.getResult().isMoveValid() ) {
                     this.errorMessage = msg;
              } else {
                    this.warningMessage = msg;
              }
        );
```
Implicit to the above, the return type of `getErrorOrWarningMessage` is `Optional<String>`. Given the above code, I wonder what you would think about splitting the error and warning conditions? I think that may help with a potential SRP violation code-smell. The setters would then be:


```
this.errorMessage = getErrorMessage(allUnitsResult).orElse(null);
this.warningMessage = getWarningMessage(allUnitsResult).orElse(null);
```

Reply:
But the way the code is set up right now is it's the same behavior?

Do you mean just having getErrorMessage() and getWarningMessage() both call getErrorOrWarningMessage() based on lastResult?

Reply:
There are multiple ways to achieve the objective here. IMO converting to an enum flag solves a number of concerns, including this one. Otherwise, your hunch to use multiple helper methods sounds reasonable.